/* Copyright 2020 The FedLearner Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
import "google/protobuf/timestamp.proto";

package fedlearner.common;

message values_normal{
  bytes values = 1;
  repeated uint32 shape = 2;
  string dtype = 3;
}

message values_paillier{
  repeated bytes values = 1;
  repeated uint32 shape=2;
}

message histogram_request_leader{
  uint32 base = 1;
  values_normal values = 2;
  values_normal binned_features = 3;
  repeated uint32 thresholds = 4;
}

message histogram_request_follower{
  uint32 base = 1;
  values_paillier values = 2;
  values_normal binned_features = 3;
  repeated uint32 thresholds = 4;
  bytes public_key = 5;
}

message histogram_leader{
  repeated values_normal hists =1;
}
message histogram_follower{
  repeated values_paillier hists =1;
}

service HistsService {
  rpc ComputeHists_leader (histogram_request_leader) returns (histogram_leader) {}
  rpc ComputeHists_follower (histogram_request_follower) returns (histogram_follower) {}
}